modules/porous_flow/src/kernels/PorousFlowAdvectiveFlux.C

```diff
@@ -0,0 +1,321 @@
+            component_ke[_i][_j][ph] += _JxW[_qp] * _coord[_qp] * darcyQpJacobian(jvar, ph);
+      }
+    }
+  }
+
+  /**
+   * Now perform the upwinding by multiplying the residuals at the upstream nodes by their mobilities
+   * The residual for the kernel is the sum over Darcy fluxes for each phase.
+   * The Darcy flux for a particular phase is
+   * R_i = int{massfrac*mobility*flux_no_mob} = int{massfrac*mobility*grad(pot)*permeability*grad(test_i)}
+   * for node i.  where int is the integral over the element.
+   * However, in fully-upwind, the first step is to take the massfrac*mobility outside the integral,
+   * which was done in the _component_re calculation above.
+   *
+   * NOTE: Physically _component_re[i][ph] is a measure of fluid of phase ph flowing out of node i.
+   * If we had left in massfrac*mobility, it would be exactly the component mass flux flowing out of node i.
+   *
+   * This leads to the definition of upwinding:
+   *
+   * If _component_re(i)[ph] is positive then we use mass_frac_i*mobility_i.  That is we use the upwind value of mobility.
+   *
+   * The final subtle thing is we must also conserve fluid mass: the total component mass flowing out of node i
+   * must be the sum of the masses flowing into the other nodes.
+  **/
+
+  Real knorm = 0;
+  for (unsigned int qp = 0; qp < _qrule->n_points(); ++qp)
+    knorm += _permeability[qp].tr();
+
+
+  /// Loop over all the phases
+  for (unsigned int ph = 0; ph < _num_phases; ++ph)
+  {
+
+    // FIRST:
+    // this is a dirty way of getting around precision loss problems
+    // and problems at steadystate where upwinding oscillates from
+    // node-to-node causing nonconvergence.
+    // The residual = int_{ele}*grad_test*k*(gradP - rho*g) = L^(d-1)*k*(gradP - rho*g), where d is dimension
+    // I assume that if one nodal P changes by P*1E-8 then this is
+    // a "negligible" change.  The residual will change by L^(d-2)*k*P*1E-8
+    // Similarly if rho*g changes by rho*g*1E-8 then this is a "negligible change"
+    // Hence the formulae below, with grad_test = 1/L
+    Real pnorm = 0.0;
+    Real gravnorm = 0.0;
+    for (unsigned int n = 0; n < num_nodes; ++n)
+    {
+      pnorm += _pp[n][ph]*_pp[n][ph];
+      gravnorm += _fluid_density_node[n][ph]*_fluid_density_node[n][ph];
+    }
+    gravnorm *= _gravity*_gravity;
+    const unsigned int dim = _mesh.dimension();
+    const Real cutoff = 1E-8*knorm*(std::sqrt(pnorm)*std::pow(_grad_test[0][0]*_grad_test[0][0], 0.5*(2.0 - dim)) + std::sqrt(gravnorm)*std::pow(_grad_test[0][0]*_grad_test[0][0], 0.5*(1.0 - dim)));
+    bool reached_steady = true;
+    for (unsigned int nodenum = 0; nodenum < num_nodes ; ++nodenum)
+    {
+      if (component_re[nodenum][ph] >= cutoff)
+      {
+        reached_steady = false;
+        break;
+      }
+    }
+
+    Real mobility;
+    Real dmobility;
+    /// Define variables used to ensure mass conservation
+    Real total_mass_out = 0;
```

Comment:
again, my preference is to include decimals in floating point literals. It makes the code more readable as you can immediately see that the variable being assigned to is a float (that is more important in places where you do not declare the variable as well of course, like below in the loop).